### DISCUSSION TITLE
Accessing both the new and old value of an attribute

### DISCUSSION THREAD

**Discussion by eithe** [NRules/NRules#325] [Q&A]
Hi, I'm new to NRules so bear with me, but I'm somehow missing how to access the previous value of an attribute. Say I want to define a rule that only should fire if Status specifically has changed from "Processing" to "Done".

```
public class MyObject
{
    public int Id { get; set; }
    public string Status { get; set; }
}
```

```
public class ObjectNameChangedRule : Rule
{
    public override void Define()
    {
        MyObject myObj = null;
        When()
            .Match(() => myObj , o => myObj.Status == "Done")
           // match oldObject.Status to "Processing" ?
    }
}
```

**Comment by mrhighstone** [comment]
Interested on this subject, too. Did you manage to find a solution to this?

**Reply by eithe** [reply]
No, afraid not. We were prototyping to perhaps start using NRules in our project, but that effort stopped, partly due to the question above. I'm guessing since no one replied that we might be thinking about this in the wrong way, but I didn't dig into it more.

**Reply by mrhighstone** [reply]
I understand. Can't imagine us to be the first thinking about this. The solution I have applied now is to inject both objects with one marked as 'dirty', but I was hoping for a more elegant solution.

The thing is that you have to identify the injected objects somehow, other than by their type.

**Comment by snikolayev** [comment]
To close the loop on this, it's not possible to access previous value of a fact. The reason for this is that it's actually pretty hard to do in a general case. For example, imagine that you insert a fact as a mutable object, then update the state of that object and update it in the rules engine. The before and after are actually the same object in memory, so they would be identical. To support this type of thing, the engine would have to somehow clone the fact.
Also, I think that when you start looking at before/after, you are really trying to do event processing (instead of comparing obj status to “processing” and “done” in the above example, think of an event ProcessingCompleted). So, adding event support to NRules would be a more natural progression here I believe.
If you are going down the path of inserting both versions, i suggest you add something like an Id field that you can join them on.

**Comment by mrhighstone** [comment]
Thank you for your extensive reply, I understand your considerations. For now, I have resolved it with a status property on a second object. Not ideal, but it will do.

**Reply by eithe** [reply]
Any chance you could share just a brief code example on how you achieved this?

**Reply by mrhighstone** [reply]
Initially I added a property on the object, to define a "source" and "target" object, so that the respective could be identified in the Match function.

However, after some thorough thinking, I found that - despite the overlap in data - the two objects were essentially different things and hence I decided to refactor the solution to different types of objects. This simplified the Match clause and removed the need of a distinguishing property. 

Morale: sometimes you have to think twice and refactor as you go along ;-)
